### ISSUE TITLE
Questions about results from with my own dataset

### ISSUE THREAD

**Issue by xhh232018** [JiahuiYu/generative_inpainting#75]
Hi, Jiahui! After an one-week training on a GTX 1080TI, I found some interesting results from my own dataset.
There are 2 kinds of images in my dataset. One is the images with clearly texture like this:
![7147](https://user-images.githubusercontent.com/35520437/42310552-491e4052-806e-11e8-8dbb-e2c8676ef00a.jpg)
![image](https://user-images.githubusercontent.com/35520437/42310620-7b10cb8e-806e-11e8-9f74-3bfdce57e24d.png)
The inpainting results of this kinds of images are semantic plausible:
![7182_ip](https://user-images.githubusercontent.com/35520437/42310729-bde7c1b0-806e-11e8-9707-71ce75e60ffe.jpg)
![demo16](https://user-images.githubusercontent.com/35520437/42310720-b88f4260-806e-11e8-939f-559316f5cd2c.jpg)
![demo17](https://user-images.githubusercontent.com/35520437/42310939-33da467c-806f-11e8-8885-774f031556c5.jpg)
![7147_ip](https://user-images.githubusercontent.com/35520437/42310940-34356476-806f-11e8-9db8-90daeb4ba1b9.jpg)
Also, there are some images like this that contains more information and structure:
![image01](https://user-images.githubusercontent.com/35520437/42311004-57f94490-806f-11e8-9de3-0be5842224d2.jpg)
However, the result of this image from my pre-trained model is quite blurry and bad:
![2_ip](https://user-images.githubusercontent.com/35520437/42311052-8747ee2c-806f-11e8-81b7-ccab3bfba7e6.jpg)
![16](https://user-images.githubusercontent.com/35520437/42311478-84774304-8070-11e8-9820-4d4dc10b98b8.jpg)
Here are my hypothesis:
1. The second kind of images is minority in my training set. Should I have to increase the ratio to 1:1? The total number of images in my dataset is 8000.
2. According to #21 and #53, I should do fine-tuning for my model. Could you please give my suggestions of changes of the hyper-parameters?
Here are the screenshots from Tensorboard:
![selection_070](https://user-images.githubusercontent.com/35520437/42311796-41fcc642-8071-11e8-8301-3f072f3553bb.png)
![selection_069](https://user-images.githubusercontent.com/35520437/42311797-4258d752-8071-11e8-8ab0-be17692e3bf4.png)
![selection_065](https://user-images.githubusercontent.com/35520437/42311798-4310fac6-8071-11e8-8733-b8eab07fd3cf.png)
![selection_064](https://user-images.githubusercontent.com/35520437/42311800-436cda12-8071-11e8-9d6b-536b37dd77c1.png)

**Comment by JiahuiYu** [comment]
Hi, first thanks for you interest in our work and share some your results. Some answers that may help:

1. The data balance is important. So it may help if you can increase the samples of the second case. You can either collect more examples or do data augmentations like random flipping/rotation/adjusting colors.
2. To fine-tune a pre-trained model, you do not have to change the hyper-parameters.
3. More data samples will help since in your case you only have 8k images. Usually I work on at least 30k images (up to 10 millions of images).

**Comment by xhh232018** [comment]
OK, I will try to create more training samples. Also, I mean that the pre-trained model is what I trained based my own dataset and your default hyper-parameters setting not the one you provided. Should I change the hyper-parameters if I want to refine it?

**Comment by JiahuiYu** [comment]
You don't need to change hyper-parameters in my understanding, unless you find some failure cases like ones addressed in issue #53 and #21.

**Comment by xhh232018** [comment]
OK. Thanks for your help. I'll try to train a new model based on more training samples ASAP and I will give you my latest results after several days.

**Comment by xhh232018** [comment]
@JiahuiYu ，Sorry to bother you again. I want to re-implement the deepfill v2 based on your deepfill v1 since deepfill v1 cannot handle irregular masked images. Thus, I want to confirm that the gated convolution  layers are only used in the coarse network. Should I have to replace the vanilla convolution layers with gated convolution layers in the refinement network? Thanks for your help.

**Comment by JiahuiYu** [comment]
Gated convolution are used in both networks. I think it is important to use gated convolution in refinement network as well.

**Comment by JiahuiYu** [comment]
@xhh232018 Hi first thanks for your interest and I saw you already carefully read the paper and code. I appreciate. For your questions:

1. The fig3 in paper shows that masks are also concatenated as input. Also I concatenated the ones. The reason is addressed in issue #40. Sorry I forget to mention concatenating ones in the paper.

2. Keep the last convolution as it is instead of gated convolution.

3. The implementation of contextual attention layer need binary mask to indicate which pixels are missing and need to be reconstructed. So your understanding is correct.

4. Actually I already released the implementation of sn-gan loss in dev branch in neuralgym. :)

**Comment by xhh232018** [comment]
@JiahuiYu Thanks a lot! You helped me a lot. I really appreciate that you can answer my questions so quickly. However, there are still some details about the discriminator that I want to confirm:
1. In your paper, you mentioned the input for the discriminator consists of image, mask and guidance channels, which is different from Deepfill V1(only the batch_pos_neg(original image and predicted image). Therefore, I want to know how to I create the input. Can I concatenate them as the input of inpainting_net like:
![image](https://user-images.githubusercontent.com/35520437/42733041-b8096108-885d-11e8-8c89-7aed6c4edb20.png) 
and then spilt it:
![image](https://user-images.githubusercontent.com/35520437/42733052-d4fccfe8-885d-11e8-916f-8bb894152110.png)
2. I notice that the flatten and fully-connected layers were used in Deepfill V1:
![image](https://user-images.githubusercontent.com/35520437/42731502-d8360bb0-8840-11e8-8e18-b24659282299.png)
![image](https://user-images.githubusercontent.com/35520437/42731503-db4183ca-8840-11e8-949b-fa02207059d3.png)
According to your latest paper, it seems that you abandoned them. Should I deleted these layers so that the input of sn-gan loss is a tensor with shape of [batchsize,16,16,256] (assume the size of input image is 512*512)or keep them?
